tensorly/backend/core.py

```diff
@@ -0,0 +1,854 @@
+        Note that this only affects threads where the backend hasn't already
+        been explicitly set. If False (default) the backend is only set
+        contextually for the local thread.
+
+    Examples
+    --------
+    Set the backend to numpy globally for this thread:
+
+    >>> import tensorly as tl
+    >>> tl.set_backend('numpy')
+
+    Set the backend to numpy globally for all threads:
+
+    >>> tl.set_backend('numpy', make_default=True)
+
+    Set the backend to use pytorch inside a context:
+
+    >>> with tl.set_backend('pytorch'):  # doctest: +SKIP
+    ...     pass
+    """
+    def __init__(self, backend, make_default=False):
+        if not isinstance(backend, Backend):
+            if backend not in _LOADED_BACKENDS:
+                # load the backend
+                if backend in _KNOWN_BACKENDS:
+                    importlib.import_module('tensorly.backend.%s_backend' % backend)
+                else:
+                    msg = ("Unknown backend %r, known backends are {%s}"
+                           % (backend, ', '.join(map(repr, _KNOWN_BACKENDS))))
+                    raise ValueError(msg)
+            backend = _LOADED_BACKENDS[backend]
+
+        # Set the backend
+        self._old_backend = getattr(_STATE, 'backend', None)
+        self._new_backend = _STATE.backend = backend
+
+        if make_default:
+            global _DEFAULT_BACKEND
+            _DEFAULT_BACKEND = backend
+
+    def __repr__(self):
+        return 'tensorly.set_backend(%r)' % self._new_backend.backend_name
+
+    def __enter__(self):
+        return None
+
+    def __exit__(self, *args):
+        if self._old_backend is None:
+            del _STATE.backend
+        else:
+            _STATE.backend = self._old_backend
+
+
+def get_backend():
+    """Returns the backend currently used
+
+    Returns
+    -------
+    backend_name : str
+        The backend currently in use
+    """
+    return getattr(_STATE, 'backend', _DEFAULT_BACKEND).backend_name
+
+
+class BackendAttributeModuleType(types.ModuleType):
+    """A module type to dispatch backend generic attributes."""
+
+    _dynamic_attributes = {'int64', 'int32', 'float64', 'float32', 'SVD_FUNS'}
+
+    def __getattr__(self, key):
+        if key in self._dynamic_attributes:
+            backend = getattr(_STATE, 'backend', _DEFAULT_BACKEND)
+            return getattr(backend, key)
+        raise AttributeError("module %r has no attribute %r"
+                             % (self.__name__, key))
+
+    def __dir__(self):
+        out = set(super(BackendAttributeModuleType, self).__dir__())
+        out.update(self._dynamic_attributes)
+        return list(out)
+
+
+def wrap_module(module_name):
+    """Wrap a module to dynamically dispatch attributes to the registry.
+
+    Intended use is
+
+    >>> registry.wrap_module(__name__)
+    """
+    sys.modules[module_name].__class__ = BackendAttributeModuleType
```

Comment:
This is also pretty neat -- but is it needed?
Unless we have a good reason to use it, i worry it ends up obfuscating needlessly what is happening -- also making understanding and maintaining the code harder, especially for new comers.

Reply:
Fair enough - as long as we document this clearly :)
What do you mean by `registry`? Could we put a clearer example using actual tensorly code?

Reply:
Oops, that docstring is from an older implementation. Can fix.